programs/account-compression/src/instructions/insert_into_indexed_array.rs

```diff
@@ -0,0 +1,152 @@
+use light_hasher::Poseidon;
+use light_indexed_merkle_tree::array::IndexingArray;
+
+use crate::RegisteredProgram;
+
+#[derive(Accounts)]
+pub struct InsertIntoIndexedArrays<'info> {
+    /// CHECK: should only be accessed by a registered program/owner/delegate.
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    pub registered_verifier_pda: Option<Account<'info, RegisteredProgram>>, // nullifiers are sent in remaining accounts. @ErrorCode::InvalidVerifier
+}
+
+/// Inserts every element into the indexed array.
+/// Throws an error if the element already exists.
+/// Expects an indexed queue account as for every index as remaining account.
+pub fn process_insert_into_indexed_arrays<'a, 'b, 'c: 'info, 'info>(
+    ctx: Context<'a, 'b, 'c, 'info, InsertIntoIndexedArrays<'info>>,
+    elements: &'a [[u8; 32]],
+    low_element_indexes: &'a [u16],
+) -> Result<()> {
+    if low_element_indexes.len() != elements.len() {
+        msg!(
+            "Number of low  does not match number elements {} != {}",
+            low_element_indexes.len(),
+            elements.len()
+        );
+        return err!(crate::errors::AccountCompressionErrorCode::NumberOfLeavesMismatch);
+    }
+    if elements.len() != ctx.remaining_accounts.len() {
+        msg!(
+            "Number of elements does not match number of indexed arrays accounts {} != {}",
+            elements.len(),
+            ctx.remaining_accounts.len()
+        );
+        return err!(crate::errors::AccountCompressionErrorCode::NumberOfLeavesMismatch);
+    }
+    // for every index
+    let mut array_map = HashMap::<Pubkey, (&'info AccountInfo, Vec<[u8; 32]>, Vec<u16>)>::new();
+    for (i, mt) in ctx.remaining_accounts.iter().enumerate() {
+        match array_map.get(&mt.key()) {
+            Some(_) => {}
+            None => {
+                array_map.insert(mt.key(), (mt, Vec::new(), Vec::new()));
+            }
+        };
+        array_map.get_mut(&mt.key()).unwrap().1.push(elements[i]);
+        array_map
+            .get_mut(&mt.key())
+            .unwrap()
+            .2
+            .push(low_element_indexes[i]);
+    }
+
+    for (mt, elements, low_element_indexes) in array_map.values() {
+        let array = AccountLoader::<IndexedArrayAccount>::try_from(mt).unwrap();
+        let mut array_account = array.load_mut()?;
+        let array = indexed_array_from_bytes_mut(&mut array_account.indexed_array);
+        for (element, _index) in elements.iter().zip(low_element_indexes) {
+            //   msg!("Inserting element {:?} into indexed array", element);
+
+            array
+                .append(
+                    // *index, TODO: enable index once we have rpc to get the low elements from indexer and a correction function inside the append function the index should just be a starting point
+                    BigInteger256::deserialize_uncompressed_unchecked(element.as_slice()).unwrap(),
+                )
+                .unwrap();
+        }
+    }
```

Comment:
yes,  because we don't have a way yet to recover the low element in case the queue state changed and the low element is invalid at execution time. This will happen a lot for sure since the queue is being emptied constantly.

Reply:
> OK, as far as I understand this code:
> 
>     * We assume that `remaining_accounts`, `elements` and `low_element_indexed` are the same - at least that's what the checks above tell me.
> 
>     * For each account (let's assume it has an index `i`), we want to append element `i` and `low_element` `i`.
> 
> 
> I don't see a point of having this nested HashMap here. I doubt that using it all together is necessary, but even if there is something I'm missing, why does it have vectors inside?
> 
> I feel like this could be just rewritten to either the classic C style, which makes sense for me here 😅

You are right we require that every remaining account  belongs to the respective element and low element.
However, the remaining accounts might contain the same Merkle tree twice.
The first loop eliminates duplicate Merkle tree accounts and stores a vector of the indices of respective leaves that will be inserted into that Merkle tree.
The second loop performs the actual inserts of all leaves of that tree so that we can take advantage of batch insertion.

I have come to not like doing this onchain though I will move this logic to the client.

Reply:
I tried to move it to the client and it didn't result in much simplification.
Do you have a better idea?
(Otherwise I will leave it like this for now and keep it in mind while building the other programs.)